docs(template): add getting-started and contracts (#1733)
latticexyz/mud: next-docs/pages/templates/typescript/contracts.mdx

# Contracts (Vanilla and React)

This section explains the contracts for the vanilla and react templates.

> The Three.JS template uses diffeerent contracts.

Reviewer: ```suggestion
The Three.JS template uses different contracts.
```

The onchain components can be divided into two types of functionality:

> - **Data**, the part of the system that stores the information.
> - **Logic**, the part of the system that can be called to read or modify the data.

Reviewer: I know you mean system in the general sense here, but i feel like it might be confusing with the "logic contracts" being called `System` as well. How about something in the direction of this:

```suggestion
- **Tables**, storing the data of the application.
- **Systems**, business logic that can be called to read or modify data in the tables.
```

## Data

### The data schema

#### `mud.config.ts`

> The data schema is declared in `packages/contracts/mud.config.ts`.

Reviewer: I feel like "table schema" would be more precise here (especially if we go with the suggestion above)

Author: Done.

> [Read more details about the schema definition here](/world/config#global-configuration-keys).

Reviewer: we'll have to update this link to the matching place in the new docs once it's ready

Reviewer: actually, let's just point it to the new place already
```suggestion
[Read more details about the schema definition here](/cli/config).
```

The data schema provided in the example is extremely simple (one singleton).

> To see more of how it works, lets add a second table:

Author: No. I decided that doesn't belong in a walk-through.

> To see more of how it works, lets add a second table:

Reviewer: makes sense, let's remove the sentence then 
```suggestion
```

```ts filename="mud.config.ts" copy
import { mudConfig } from "@latticexyz/world/register";

export default mudConfig({
  tables: {
    Counter: {
      keySchema: {},
      valueSchema: "uint32",
    },
  },
});
```

There are two automatically generated files related to the data:

- `packages/contracts/src/codegen/index.sol`
- `packages/contracts/src/codegen/tables/Counter.sol`

<details>

<summary>The automatically generated data files</summary>

`index.sol`

This file just imports all of the automatically generated tables and their identifiers.

```solidity filename="index.sol"
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.21;

/* Autogenerated file. Do not edit manually. */

import { Counter, CounterTableId } from "./tables/Counter.sol";
```

In this case there is only one table, `Counter`.

`Counter.sol`

```solidity
// SPDX-License-Identifier: MIT

> pragma solidity >=0.8.0;

Reviewer: nit (does it actually generate 0.8.0? if so there might be an issue with the codegen script)
```suggestion
pragma solidity >=0.8.21;
```

Author: You're right. My bad - I updated most of the code, but forgot that line.

/* Autogenerated file. Do not edit manually. */

// Import schema type
import { SchemaType } from "@latticexyz/schema-type/src/solidity/SchemaType.sol";

> // Import store internals

Reviewer: ```suggestion
```

```solidity
/**
 * @notice Tightly pack static (fixed length) data using this table's schema.
 * @return The static data, encoded into a sequence of bytes.
 */
function encodeStatic(uint32 value) internal pure returns (bytes memory) {
  return abi.encodePacked(value);
}

/**
 * @notice Encode all of a record's fields.
 * @return The static (fixed length) data, encoded into a sequence of bytes.
 * @return The lengths of the dynamic fields (packed into a single bytes32 value).
 * @return The dyanmic (variable length) data, encoded into a sequence of bytes.
 */
function encode(uint32 value) internal pure returns (bytes memory, PackedCounter, bytes memory) {
  bytes memory _staticData = encodeStatic(value);

  PackedCounter _encodedLengths;
  bytes memory _dynamicData;

  return (_staticData, _encodedLengths, _dynamicData);
}

/**
 * @notice Encode keys as a bytes32 array using this table's field layout.
 */
function encodeKeyTuple() internal pure returns (bytes32[] memory) {
  bytes32[] memory _keyTuple = new bytes32[](0);

  return _keyTuple;
}
```

Utility functions to encode a value.

</details>

## Logic

> The way mud works, onchain logic is implemented by one or more `System`s.

Reviewer: also just noticed, let's capitalize `MUD` everywhere in text

> The way mud works, onchain logic is implemented by one or more `System`s.
> Those systems are typically called by a central `World`,

Reviewer: ```suggestion
The way mud works, onchain logic is implemented by one or more `System` contracts.
Those systems are always called by a central `World` contract.
```
